### DISCUSSION TITLE
Why is params optional in GetStaticPropsContext type?

### DISCUSSION THREAD

**Discussion by tavoyne** [vercel/next.js#35742] [Help] [answered]
This is from [https://github.com/vercel/next.js/blob/.../packages/next/types/index.d.ts](https://github.com/vercel/next.js/blob/937ab16b780dc2bc755fa9c032ffe498c79b5af6/packages/next/types/index.d.ts):

```ts
export type GetStaticPropsContext<
  Q extends ParsedUrlQuery = ParsedUrlQuery,
  D extends PreviewData = PreviewData
> = {
  params?: Q
  preview?: boolean
  previewData?: D
  locale?: string
  locales?: string[]
  defaultLocale?: string
}
```

Why is `params` optional? Correct me if I'm wrong, but it will always be provided to `getStaticProps()`. 

Right now we're forced to do a non-null assertion in order for it to be correctly typed, which is annoying, why not making it non-optional?

**Comment by tavoyne** [comment]
The only risk is to forget to return it from `getStaticPaths()`. What about having a `InferGetStaticPaths` in this case?

**Comment by icyJoseph** [comment, marked as answer]
There's one use case where a path looks like `[[...id]]`, and because we cannot bind TypeScript to the file name structure, we need to take the type that most correctly represents all use cases.

`post/[[...id]]` matches:

- post
- post/a
- post/a/b and so on.

So the `id` list could be `undefined`, and that's what the type represents.

**Reply by tavoyne** [reply]
Thanks for clarifying. Makes sense.
